### DISCUSSION TITLE
Can't communicate with Jetson Xavier NX

### DISCUSSION THREAD

**Discussion by marcelino-pensa** [iRobotEducation/create3_docs#59] [Questions]
I have a Nvidia Jetson Xavier NX, and I flashed it using [Nvidia's resource website](https://developer.nvidia.com/embedded/learn/get-started-jetson-xavier-nx-devkit).
In sequence, I hooked up the Jetson to an iRobot create using the connection as described in [this link](https://iroboteducation.github.io/create3_docs/hw/hookup/).

After that, I did what's described in [this link](https://iroboteducation.github.io/create3_docs/setup/jetson/), which would be simple enough. However, I'm not getting the expected output with `ifconfig l4tbr0`. The expected output would be:
```
l4tbr0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.186.3  netmask 255.255.255.0  broadcast 192.168.186.255
        inet6 fe80::1  prefixlen 128  scopeid 0x20<link>
        inet6 fe80::ecb6:edff:feac:7dd5  prefixlen 64  scopeid 0x20<link>
        ether ee:b6:ed:ac:7d:d5  txqueuelen 1000  (Ethernet)
        RX packets 169  bytes 36206 (36.2 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1644  bytes 213306 (213.3 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

In turn, I am getting the following:
```
l4tbr0: flags=4098<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        ether 72:18:0e:4c:ca:75  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

Any directions on what I've done wrong? I can't get to communicate with the robot at all...

**Comment by marcelino-pensa** [comment]
For reference, these are the files that I was supposed to change:

`/opt/nvidia/l4t-usb-device-mode/IP_ADDRESS_FOR_CREATE3_ROBOT.conf`:
```
192.168.186.3
```

`/opt/nvidia/l4t-usb-device-mode/nv-l4t-usb-device-mode-runtime-start.sh`:
```
#!/bin/bash

# Copyright (c) 2017-2020, NVIDIA CORPORATION.  All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

script_dir="$(cd "$(dirname "$0")" && pwd)"
. "${script_dir}/nv-l4t-usb-device-mode-config.sh"

if [ -f "${script_dir}/IP_ADDRESS_FOR_CREATE3_ROBOT.conf" ]; then
    robot_ip_conf_file=${script_dir}/IP_ADDRESS_FOR_CREATE3_ROBOT.conf
    cat $robot_ip_conf_file
fi

/sbin/ifconfig l4tbr0 up
# The interface might lose the address when down. Add it here to make sur
# it's configured.
/sbin/ifconfig l4tbr0 ${net_ip} netmask ${net_mask}
/sbin/ifconfig l4tbr0 add ${net_ipv6}

# IF the config file for iRobot Create® 3 (IP_ADDRESS_FOR_CREATE3_ROBOT.conf) exists, 
# then apply the static IP address (192.168.186.3), instead of enabling DHCP server
if [ -n "$robot_ip_conf_file" ]; then
    static_ip_for_robot=`cat $robot_ip_conf_file`
    /sbin/ifconfig l4tbr0 $static_ip_for_robot netmask 255.255.255.0 broadcast 192.168.186.255
    echo "Static IP address set to $static_ip_for_robot"
else

# Start a DHCP server so that connected systems automatically receive an IP
# address. This avoids users having to manually configure the connection, and
# also prevents Network Manager on Linux from destroying any manually applied
# configuration.
#
# The DHCP server must be started here, because it won't start if it's told to
# run on an interface that's down.
if [ -n "${net_dhcp_start}" ]; then
    echo "### IN"
    dhcpd_conf="${script_dir}/dhcpd.conf"
    dhcpd_leases="/run/l4t-usb-devmode-dhcpd.leases"
    dhcpd_pid="/run/l4t-usb-devmode-dhcpd.pid"
    cat > "${dhcpd_conf}" <<ENDOFHERE
max-lease-time ${net_dhcp_lease_time};
default-lease-time ${net_dhcp_lease_time};

subnet ${net_net} netmask ${net_mask} {
    range ${net_dhcp_start} ${net_dhcp_end};
}
ENDOFHERE
    rm -f "${dhcpd_leases}"
    touch "${dhcpd_leases}"
    /usr/sbin/dhcpd \
        -cf "${dhcpd_conf}" \
        -pf "${dhcpd_pid}" \
        -lf "${dhcpd_leases}" \
        l4tbr0
fi

# When an interface is set to "down", all routes through it are lost, so we
# must restore any routes when setting the interface to "up".
if [ -n "${net_ipv4_defroute_router}" ]; then
    route add default gw ${net_ipv4_defroute_router} dev l4tbr0 \
        ${net_ipv4_defroute_metric:+metric ${net_ipv4_defroute_metric}}
fi

fi

exit 0
```

**Reply by shamlian** [reply]
The Jetson document was written by our friend @tokk-nv at NVIDIA; perhaps he can assist?

**Comment by shamlian** [comment]
One suspicion: have you switched the USB/BLE toggle on the [adapter board](https://iroboteducation.github.io/create3_docs/hw/adapter/#usbble-toggle) to "USB?"

**Reply by marcelino-pensa** [reply]
Ooh, thanks for asking that, as I didn't specify in the OP... 
but yeah, I did make sure that the switch is on the USB side, see the image below

![IMG_20220517_181119](https://user-images.githubusercontent.com/51488981/168926910-d93d6f95-49ba-4154-bf54-1cd0131db8fd.jpg)

**Reply by justinIRBT** [reply]
@marcelino-pensa I don't have a Jetson handy to try to reproduce, but just to get you unblocked, are you able to connect both the Create3 and the Jetson to your wifi?  They should be able to communicate with each other over wifi until you get the usb configuration resolved.  You can use the `Phase 2: Update Robot` instructions here https://edu.irobot.com/create3-setup through Step 4 to get on the robot's web app.  From there you can use the `Connect` tab on the robot's web app to get your robot connected to a 2.4 GHz network.

While on the robot's web app, there is one piece of information we can get to make sure the Create3 sees the Jetson over USB.  If you go the `Logs` tab on the robot's web app, then type `usb` into the dialogue next to `Search Logs` and click the button, if the Create3 registered that is had a usb connection you should see the log prints:
```
May 18 19:00:13 Create+EP2 daemon.info avahi-daemon[1203]: Joining mDNS multicast group on interface usb0.IPv4 with address 192.168.186.2.
May 18 19:00:13 Create+EP2 daemon.info avahi-daemon[1203]: New relevant interface usb0.IPv4 for mDNS.
May 18 19:00:13 Create+EP2 daemon.info avahi-daemon[1203]: Registering new address record for 192.168.186.2 on usb0.IPv4.
```
![LogsPic](https://user-images.githubusercontent.com/69175069/169138859-0e92f0b7-98b3-42e3-a304-e734977b8eb8.png)

**Reply by marcelino-pensa** [reply]
@justinIRBT Thanks for your answer... see below the logs after I did what you told me to:
```
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.402436] usb 1-1: new high-speed USB device number 12 using ehci-platform
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.602864] usb 1-1: New USB device found, idVendor=0955, idProduct=7020
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.602908] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.602917] usb 1-1: Product: Linux for Tegra
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.602928] usb 1-1: Manufacturer: NVIDIA
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.602939] usb 1-1: SerialNumber: 1420821062636
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.616047] usb-storage 1-1:1.4: USB Mass Storage device detected
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.647216] scsi host0: usb-storage 1-1:1.4
May 18 19:43:38 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85113.679838] cdc_ncm 1-1:1.5 usb0: register 'cdc_ncm' at usb-101c0000.ehci-1, CDC NCM, 72:18:0e:4c:ca:76
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 daemon.info avahi-daemon[12522]: Joining mDNS multicast group on interface usb0.IPv4 with address 192.168.186.2.
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 daemon.info avahi-daemon[12522]: New relevant interface usb0.IPv4 for mDNS.
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 daemon.info avahi-daemon[12522]: Registering new address record for 192.168.186.2 on usb0.IPv4.
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85114.339047] cdc_ncm 1-1:1.5 usb0: 425 mbit/s downlink 425 mbit/s uplink
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85114.371044] cdc_ncm 1-1:1.5 usb0: 425 mbit/s downlink 425 mbit/s uplink
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 kern.info kernel: [85114.403033] cdc_ncm 1-1:1.5 usb0: network connection: connected
May 18 19:43:39 iRobot-A89651CF962B48E085A768E81366DAD4 kern.err kernel: [85114.403767] cdc_ncm 1-1:1.5 usb0: kevent 2 may have been dropped
```

As for connecting both to my wifi, I was able to switch the network on the Create3, but then I couldn't access the web browser tool through `192.168.10.1`. I was able to `ping iRobot.local`, so I knew it was connected to my wifi throught the IP `192.168.0.13` (possibly assigned through DHCP). I didn't know what to do from there

**Reply by marcelino-pensa** [reply]
Here are some more relevant logs. I started the docker container and I can ping the robot. However, I don't get much in `topic list`:
```
root@irobot-xavier:/# ping 192.168.0.13
PING 192.168.0.13 (192.168.0.13): 56 data bytes
64 bytes from 192.168.0.13: icmp_seq=0 ttl=64 time=1.976 ms
64 bytes from 192.168.0.13: icmp_seq=1 ttl=64 time=1.667 ms
64 bytes from 192.168.0.13: icmp_seq=2 ttl=64 time=1.664 ms
^C--- 192.168.0.13 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max/stddev = 1.664/1.769/1.976/0.146 ms

root@irobot-xavier:/# ros2 topic list
/parameter_events
/rosout
```

**Reply by marcelino-pensa** [reply]
I just tried to play with the param `CYCLONEDDS_URI`, but no success yet
The Jetson is connected via USB, but nothing comes if I use `l4tbr0` as the network interface:
```
root@irobot-xavier:/# ros2 topic list
/parameter_events
/rosout

root@irobot-xavier:/# echo $CYCLONEDDS_URI
<CycloneDDS><Domain><General><NetworkInterfaceAddress>l4tbr0</NetworkInterfaceAddress></General></Domain></CycloneDDS>
```

Then I tried to change the network interface to the known IP, but still the same issue:
```
root@irobot-xavier:/# echo $CYCLONEDDS_URI
<CycloneDDS><Domain><General><NetworkInterfaceAddress>192.168.0.13</NetworkInterfaceAddress></General></Domain></CycloneDDS>

root@irobot-xavier:/# ros2 topic list
/parameter_events
/rosout

root@irobot-xavier:/# ping 192.168.0.13
PING 192.168.0.13 (192.168.0.13): 56 data bytes
64 bytes from 192.168.0.13: icmp_seq=0 ttl=64 time=1.859 ms
```

This is my first experience with ros2 despite tons of experience with ROS1... all this is new territory to me, so any help is appreciated

**Reply by marcelino-pensa** [reply]
I added a third machine to the network and launched `ros2 run demo_nodes talker`... then I was able to verify that the jetson could see ROS2 messages coming in from that machine, but nothing from the Create3:
```
root@irobot-xavier:/# export CYCLONEDDS_URI='<CycloneDDS><Domain><General><NetworkInterfaceAddress>wlan0</NetworkInterfaceAddress></General></Domain></CycloneDDS>'

root@irobot-xavier:/# ros2 topic echo /chatter
data: 'Hello World: 54'
---
data: 'Hello World: 55'
---
data: 'Hello World: 56'
---

root@irobot-xavier:/# ros2 topic list
/chatter
/parameter_events
/rosout
```

**Reply by justinIRBT** [reply]
Once the create3 is on the wifi, you can access its web application from its ip address on the network.  In this case, if you put `192.168.0.13` into a browser, you should land on the robot's web page.

Regarding not seeing topics, first off, can you try running:
```
ros2 topic list --no-daemon --spin-time 10
```
A little explanation for trying that command is here (https://iroboteducation.github.io/create3_docs/api/ros2/)

If that doesn't work, can you first try using FastRTPS, it handles multi-interface networking a little better.  We can at least confirm you can see the topics that way, then look into cyclone.  First, use the robot web app to set the Create3 to FastRTPS.  If you go to the `Application` tab on the robot's web browser, you can select the RMW implementation (https://iroboteducation.github.io/create3_docs/setup/provision/).  Select fastrtps, then restart the application.

After your robot makes the happy sounds, on your Jetson terminal, run:
```
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
ros2 topic list --no-daemon --spin-time 10
```

Lets see if that will get you a list of the topics, if it does and you would prefer to use Cyclone, you can set the robot back to cyclone and look into some of the config options described here (https://iroboteducation.github.io/create3_docs/setup/xml-config/).

**Reply by marcelino-pensa** [reply]
@justinIRBT thank you for your support! I can communicate with the robot using wifi right now... as reference for other users, I made sure that Create3 was in the same network as my compute board, then I needed the following configurations in the docker container:
```
export CYCLONEDDS_URI='<CycloneDDS><Domain><General><NetworkInterfaceAddress>wlan0</NetworkInterfaceAddress></General></Domain></CycloneDDS>'
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
```

Then I would get data just with the command:
```
ros2 topic list --no-daemon
```

I could not make it work using USB, there seems to be an issue with `l4tbr0`, but thanks again since you've helped me to get started. I'll keep the discussion open until we can figure out the USB connection issue

**Reply by alsora** [reply]
@marcelino-pensa I'm happy it works for you now!
Just FYI, when you set the `RMW_IMPLEMENTATION=rmw_fastrtps_cpp` you are switching RMW implementation from CycloneDDS to Fast-DDS.
For this reason you don't need to export a `CYCLONEDDS_URI` as it is ignored by the application.
I recommend you to read this documentation page to find more information about ROS 2 middlewares: https://iroboteducation.github.io/create3_docs/setup/xml-config/

**Reply by justinIRBT** [reply]
@marcelino-pensa Something else to try to get the USB working, can you try using `nmtui` to configure the usb.  

Use `sudo nmtui`  to add a network connection.  Add a connection of type Ethernet with a manual IP address of 192.168.186.1/24 - the network connection should look like this in nmtui (your USB address and device MAC Address may vary):
![Screenshot from 2022-05-18 18-26-25](https://user-images.githubusercontent.com/69175069/169182352-9a19029e-d34e-498c-b5e7-42fbbffdc1bd.png)
I have these instructions from another system setup (not Create3) and don't have a Create3 + Jetson to test on, so I am not sure if this will work for you or not, but it might be worth a quick test.  If it works, you should be able to ping 192.168.186.2 (the robot's USB IP)

**Reply by marcelino-pensa** [reply]
@justinIRBT  I had a weird experience with this...
I did as you mentioned, added the MAC address and the IP you suggested, and then I executed `ifconfig l4tbr0`, and the following showed up:
```
l4tbr0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.55.1  netmask 255.255.255.0  broadcast 192.168.55.255
        ether 72:18:0e:4c:ca:75  txqueuelen 1000  (Ethernet)
        RX packets 24  bytes 3108 (3.1 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 774  bytes 194777 (194.7 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

Now at least I have a connection and I can ping `192.168.55.1`. The weird thing is that this is not the IP I chose initially. Then I opened `nmtui` again, and I could see that there was a Bridge created called `l4tbr0`. Not sure how it showed up, as I didn't create it:
![image](https://user-images.githubusercontent.com/51488981/169409466-6aaf3486-2090-43b4-be01-a1ddce8659e5.png)

On top of that, despite being able to ping `192.168.55.1`, I wasn't able to get any ROS data if the robot and I are in different wifi connections. This indicates to me that I'm not being able to get data through USB, just through wifi

Anyway, thanks for pointing out this possible solution

**Reply by justinIRBT** [reply]
Why did you do `192.168.55.1` for your address, you should do `192.168.186.1` to put it on the same subnet as the robot's `192.168.186.2`.  Re-reading your comment, it looks like you set 192.168.186.1, but it was changed with that config.  The USB1 from above was the wrong address (that was from a system where the USB was plugged into one of the USB-A ports).  Are you able to manually change the l4tbr0 ip to 192.168.186.1/24?

**Comment by martinerk0** [comment]
I also have similar problem with create3 + Orin Nano.
On Orin Nano I have Ubuntu 20.04 and Foxy.

When using wifi, I get data with `ros2 topic echo /kidnap_status`, but when trying to use l4tbr0 with USB cable I get nothing.
I followed this [guide](https://iroboteducation.github.io/create3_docs/setup/jetson/), and with USB I can ping ping 192.168.186.2 and also see website and see connected cable in logs.
But when I try to echo messages or 
```
ros2 topic list --no-daemon --spin-time 10
1703678174.920930 [32]       ros2: using network interface l4tbr0 (udp/192.168.186.3) selected arbitrarily from: l4tbr0, docker0
/parameter_events
/rosout

```
@justinIRBT This thread is over year old, have you tried create3 with any Jetson over USB since?
I might get newer ROS version with Jetpack 6, but it would be nice to get it working with Foxy and 20.04 too

**Reply by martinerk0** [reply]
I updated the create3 firmware to the  latest 
```
>>>>>>> system version
G.5.3
```
and now the USB only communication works, although only with  `ros2 topic list --no-daemon` but I think that's because I'm using foxy instead of newer ROS release

**Reply by slowrunner** [reply]
@martinerk0 Create3 with ROS 2 Galactic firmware will only work properly with a host using ROS 2 Galactic.  

Mixing Galactic with Foxy may appear to work at first glance, but most of the Create3 functions will not work properly.  

BTW, Foxy reached End Of Life in June 2023, and Galactic was EOL in Dec 2022. Moving to Humble will ensure support until May 2027.

Also, it is best to start a new question with your specific issue.
